src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPool.cs

```diff
@@ -2,130 +2,440 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System.Collections.Concurrent;
 using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
 
 namespace System.Net.Http
 {
+    /// <summary>Provides a pool of connections to the same endpoint.</summary>
     internal sealed class HttpConnectionPool : IDisposable
     {
-        private readonly ConcurrentConnectionSet _activeConnections;
-        private readonly ConcurrentBag<HttpConnection> _idleConnections;
-        private bool _disposed;
+        /// <summary>Maximum number of milliseconds a connection is allowed to be idle in the pool before we remove it.</summary>
+        private const int MaxIdleTimeMilliseconds = 100_000;
+        /// <summary>Value used with <see cref="_associatedConnectionCount"/> to indicate that no connection has been added to the new pool yet.</summary>
+        private const int NewPoolSentinel = int.MinValue;
```

Comment:
Technically there's still a race condition if you do that, if e.g. two cleanups happened between the time the pool was created and the connection was added to it.  Since connections could take a while to be created, it didn't seem entirely far fetched that could happen.  But if you think it's overkill, I can remove it.